docs/design/2021-10-15-analyze-predicate-columns.md

```diff
@@ -0,0 +1,73 @@
+```sql
+CREATE TABLE IF NOT EXISTS mysql.column_stats_usage (
+	table_id BIGINT(64) NOT NULL,
+	column_id BIGINT(64) NOT NULL,
+	last_used_at TIMESTAMP,
+	last_analyzed_at TIMESTAMP,
+	PRIMARY KEY (table_id, column_id) CLUSTERED
+);
+```
+`table_id` and `column_id` are the IDs of the table and the column respectively. `last_used_at` is the timestamp at which the column stats were used(needed). `last_analyzed_at` is the timestamp at which the column stats are updated.
+
+During the plan optimization phase, we track which column stats are used(needed) in memory. `updateStatsWorker` periodically dumps column stats usage(in memory) into `mysql.column_stats_usage`(in disk). Since tracking predicate columns in memory is a high frequency operation, we provide a switch to control whether to enable tracking predicate columns for insurance.
+
+Each time when executing `ANALYZE PREDICATE COLUMNS`, we read predicate columns(the columns whose `last_used_at` is not null) from `mysql.column_stats_usage` and then decide which columns' statistics need to be collected.
+
+If a table or a column is deleted, the related rows in `mysql.column_stats_usage` also need to be deleted.
+
+Besides, we provide the syntax `show column_stats_usage` to let users see which columns' stats are used(needed) or updated.
+
+### Auto Analyze Logic
+
+By default, auto-analyze would collect the statistics of all columns. We can modify the `ANALYZE` configuration for a specific table. The configuration includes the column option(all columns/predicate columns/user-specified columns). Once we set the configuration for the table, auto-analyze can decide which columns' statistics need to be collected according to the column option.
+
+### Modify Count And Outdated Statistics
+
+`modify_count` is currently at the table level. Each time when the table is analyzed, `modify_count` is set to 0. After introducing the `PREDICATE COLUMNS`/`COLUMNS ColumnNameList` option for `ANALYZE`, some columns' statistics are updated while others' statistics are not(maybe even not exist), but we still set `modify_count` to 0 after `ANALYZE` is finished, which breaks the table level meaning of `modify_count`. A method is to make each column have its own `modify_count` but we don't consider that currently since it involves many logic changes. Another method is to delete the outdated statistics of the columns which are excluded in the current `ANALYZE` statement but we don't adopt that since deleting the outdated statistics may bring the risk of changing plans. Therefore, our behavior is to set `modify_count` to 0 and remain outdated statistics, though it breaks the table level meaning of `modify_count`. It should be emphasized that users should list **all the columns whose statistics need to be collected** rather than parts of them, and TiDB will give a note-level warning to the client to address that each time when `ANALYZE COLUMNS ColumnNameList` is executed. 
```

Comment:
Do we have a plan to improve `modify_count` feature? IMO, setting `modify_count` to 0 and remaining outdated statistics break the definition of `modify_count` and are risky.

Reply:
I agree that it is wired that `modify_count` is about the **whole** table while we can analyze **some** of columns. We don't make each column/index have its own `modify_count` because the following two reasons:
1. It involves lots of logic changes.
2. https://docs.google.com/document/d/1VjJjAkp_EzUBroOzkZHq9OszrrY-1gnMxT1N_KYFpZs/edit?disco=AAAAIYi-D-M

From another perspective, if we set `modify_count` to 0 and delete outdated statistics, do we break the definition of `modify_count`? If the answer is no and we regard outdated statistics as better pseudo statistics, setting `modify_count` and remaining outdated statistics seems to make more sense.

The real risk of breaking the definition of `modify_count` is that we use `modfiy_count/count` to decide whether we need to analyze the table. After `ANALYZE PREDICATE COLUMNS/COLUMNS c1, ..., cn`, `modify_count` is set to 0. Then maybe some queries need statistics of a certain column, which are not collected in last analyze and the outdated statistics are used. The worrying thing is that now `modify_count/count` is 0 and neither auto-analyze nor manual-analyze will be executed on the table(the user checks stats-health and finds it 100% healthy). Another similar case is that after `ANALYZE`, `modify_count` is set to 0 and later a new index is added. In this case, auto-analyze will be triggered by the newly added index while manual-analyze may be not(the user checks stats-health and finds it 100% healthy and may forget the statistics of the newly added index have not been collected yet). 

Therefore, I think we can add another dimension to decide whether we need to analyze the table(i.e., whether the table is healthy) in addition to `modify_count/count`. When outdated/pseudo statistics of the column/index are used, we record that and use a syntax such as `show column_or_index_needed_to_be_analyzed` to show users the statistics of some columns/indexes need to be collected even though `modify_count/count` is very small or even 0.
Besides, once outdated statistics of columns are used by optimizer, it can trigger auto-analyze(just like the newly added index can trigger auto-analyze).